### ISSUE TITLE
ARM64 - sorry to open this as an issue

### ISSUE THREAD

**Issue by sm-moshi** [InBetweenNames/gentooLTO#738]
Sorry, for opening this, but I didn't know how to address this.

Does it make sense to use LTO for an ARM64 Raspberry Pi on Gentoo?

**Comment by InBetweenNames** [comment]
Sure, but you might consider cross compiling instead of building on the RPi itself.  You can use `distcc` to do that fairly easily.

**Comment by AnonymousRetard** [comment]
Unfortunately I don't think you get any speedups when using distcc with LTO though. It seems like the linking step runs locally and it probably needs to. Otherwise I would assume that every computer part of the distcc cluster needs to have exactly the same libraries.

With LTO most of the compilation time happens during the linking step.
A pull request to distcc has recently been accepted that stops it from even trying to distribute LTO work:
https://github.com/distcc/distcc/pull/413

Personally I had been wondering for a long time why distcc isn't helping my systems anymore and it's probably because of this. Most of my packages actually seem to compile a bit slower and distribute very badly when trying to use an older version of distcc.

Since I have LTO enabled system-wide except when I build the kernel it's only the kernel that I see big speed improvements on when using distcc.

**Comment by shelterx** [comment]
Yes, the linking is done locally but it also depends on the source code, bigger packages certainly compiles faster using distcc.
llvm went down to ~40 mins from 1h30mins.
The local machine is an i5 laptop, the distcc "server" is an i7 2600k.

**Comment by AnonymousRetard** [comment]
@shelterx Are you sure you actually built llvm with -flto though?
Because of this issue: https://github.com/InBetweenNames/gentooLTO/issues/619 -flto is actually stripped from the llvm package in this overlay:
`/etc/portage/package.cflags/lto.conf:sys-devel/llvm *FLAGS-=-flto* # Issue #619 temporarily disabled for now due to build errors`

**Comment by shelterx** [comment]
@AnonymousRetard Ooops, you are correct and that would explain why I don't see some stuff getting passed to distcc server.
However qtcore is compiled with -flto=auto and it's faster but I agree, it doesn't help THAT much but overall I think you gain more than you lose.
no distcc:
2021-04-13T15:19:03 >>> dev-qt/qtcore: 7′16″
distcc:
2021-05-11T15:14:36 >>> dev-qt/qtcore: 5′39″

here's another example:
no distcc:
2021-05-03T12:11:13 >>> kde-apps/kate: 3′35″
distcc:
2021-05-14T10:37:13 >>> kde-apps/kate: 2′23″

**Comment by AnonymousRetard** [comment]
@shelterx This is quite interesting. I might do some of my own tests on these packages later. I have a 4 core weak AMD system used as a server and a strong 16 core 5950X. I don't have specific examples since it's a long time ago I tried this last but I remember being very disappointed in DISTCC performance and actually seeing slowdowns from it on quite a few packages. Very few jobs where being distributed to the 5950X and the majority of the time when building packages was spent compiling things locally. These issues completely disappeared when building packages without -flto but I decided that I rather build stuff locally with LTO than try to speed up the jobs with DISTCC.

This discussion should perhaps continue somewhere else though. The issue tracker on distcc is probably a better place: https://github.com/distcc/distcc

As I mentioned in my original reply a PR has been merged that looks like distcc will soon stop trying to distribute -flto jobs completely so we'll have to raise an issue there if we want to change that behavior in a future release. Perhaps it helps in some cases but not others but I'm not sure if that can be detected automatically.
